### ISSUE TITLE
Laravel 7 + Route Cache = Login Route Problem 

### ISSUE THREAD

**Issue by davi5e** [laravel/nova-issues#2410]
- Laravel Version: 7.0.2
- Nova Version: 3.0.1
- PHP Version: 7.3

### Description:

Yesterday Laravel 7 was launched and I updated it and Nova. Once I ran my scripts to deploy the website, I noticed the `/login` route keeps redirecting to itself (sending `302`) and the culprit seems to be Artisan's cache system, which may be a bug in either Nova or Illuminate itself since the routing system was refactored...

If the following steps are not enough to reproduce the problem, I can try to debug it further provided some instructions on how to proceed.

For simplicity, **everything works without** the route caching.

### Steps To Reproduce:

- **config/nova.php**:
```php
    // default configurations hidden

    'domain' => 'admin.mydomain.com',
    'path' => '/',
```

- **app/Providers/RouteServiceProvider.php**:
```php
    public function map()
    {
        $this->mapApiRoutes();

        $this->mapAdminRoutes();

        $this->mapWebRoutes();
    }

    protected function mapAdminRoutes()
    {
        Route::domain('admin.mydomain.com')
            ->middleware('web')
            ->namespace($this->namespace)
            ->group(base_path('routes/admin.php'));
    }
```

- **routes/admin.php**:
```php
use Illuminate\Support\Facades\Route;

Route::permanentRedirect('/', '/login')
    ->name('login');
```

- Laravel Routes: No `GET /login` in sight...
```
+--------------------+----------------------------------------+-------+------------+-----------------------------------------------------+-----------------+
| Domain             | Method                                 | URI   | Name       | Action                                              | Middleware      |
+--------------------+----------------------------------------+-------+------------+-----------------------------------------------------+-----------------+
| admin.mydomain.com | GET|HEAD|POST|PUT|PATCH|DELETE|OPTIONS | /     | login      | Illuminate\Routing\RedirectController               | web             |
| admin.mydomain.com | POST                                   | login | nova.login | Laravel\Nova\Http\Controllers\LoginController@login | web,nova.guest: |
+--------------------+----------------------------------------+-------+------------+-----------------------------------------------------+-----------------+

```

- I serve the website using Docker, but it's mainly taken from [Docker for PHP Developers](https://laravel-news.com/docker-for-php-developers) to generate the containers, the only difference being `php /srv/artisan route:cache` now generates the following log:
```
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:45 +0000] 302 "GET /login" (227.844ms|mem:8Mb|cpu:96.56%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (24.664ms|mem:4Mb|cpu:40.54%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (34.902ms|mem:4Mb|cpu:57.30%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (30.184ms|mem:4Mb|cpu:66.26%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (34.424ms|mem:4Mb|cpu:87.15%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (24.139ms|mem:4Mb|cpu:82.85%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.249ms|mem:4Mb|cpu:43.01%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.569ms|mem:4Mb|cpu:84.86%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (24.849ms|mem:4Mb|cpu:80.49%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.671ms|mem:4Mb|cpu:42.25%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (22.413ms|mem:4Mb|cpu:44.62%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.882ms|mem:4Mb|cpu:41.87%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.508ms|mem:4Mb|cpu:42.54%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (24.109ms|mem:4Mb|cpu:82.96%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (27.263ms|mem:4Mb|cpu:73.36%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (34.476ms|mem:4Mb|cpu:58.01%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (24.414ms|mem:4Mb|cpu:81.92%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:46 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:46 +0000] 302 "GET /login" (23.949ms|mem:4Mb|cpu:83.51%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:47 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:47 +0000] 302 "GET /login" (24.849ms|mem:4Mb|cpu:80.49%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:47 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:47 +0000] 302 "GET /login" (25.366ms|mem:4Mb|cpu:78.85%)
nginx | 172.20.0.1 [04/Mar/2020:12:11:47 +0000] 302 "GET /login HTTP/1.1" 410 "-" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0" "-"
php-fpm     | 172.20.0.4 [04/Mar/2020:12:11:47 +0000] 302 "GET /login" (25.292ms|mem:4Mb|cpu:79.08%)
```

**Comment by rompetomp** [comment]
I also have this issue. FWIW, this is not only on subdomain routing. I also have this with the `/nova` link on the default domain

**Comment by nabeelahmeddaraz** [comment]
Happening with me too!

**Comment by crynobone** [comment]
Laravel Nova currently broken if we use `artisan route:cache`:

```
art route:list --columns=method,uri,name --name=nova
+----------+--------------------------------------------------------------------------------------------+-----------------------------+
| Method   | URI                                                                                        | Name                        |
+----------+--------------------------------------------------------------------------------------------+-----------------------------+
| GET|HEAD | login                                                                                      | nova.                       |
| POST     | login                                                                                      | nova.login                  |
| GET|HEAD | logout                                                                                     | nova.logout                 |
| GET|HEAD | nova-api/cards                                                                             | nova.api.                   |
| GET|HEAD | nova-api/dashboards/{dashboard}                                                            | nova.api.                   |
| GET|HEAD | nova-api/metrics                                                                           | nova.api.                   |
| GET|HEAD | nova-api/metrics/{metric}                                                                  | nova.api.                   |
| GET|HEAD | nova-api/scripts/{script}                                                                  | nova.api.                   |
| GET|HEAD | nova-api/search                                                                            | nova.api.                   |
| GET|HEAD | nova-api/styles/{style}                                                                    | nova.api.                   |
| GET|HEAD | nova-api/{resource}                                                                        | nova.api.                   |
| DELETE   | nova-api/{resource}                                                                        | nova.api.                   |
| POST     | nova-api/{resource}                                                                        | nova.api.                   |
| POST     | nova-api/{resource}/action                                                                 | nova.api.                   |
| GET|HEAD | nova-api/{resource}/actions                                                                | nova.api.                   |
| GET|HEAD | nova-api/{resource}/associatable/{field}                                                   | nova.api.                   |
| GET|HEAD | nova-api/{resource}/cards                                                                  | nova.api.                   |
| GET|HEAD | nova-api/{resource}/count                                                                  | nova.api.                   |
| GET|HEAD | nova-api/{resource}/creation-fields                                                        | nova.api.                   |
| GET|HEAD | nova-api/{resource}/creation-pivot-fields/{relatedResource}                                | nova.api.                   |
| DELETE   | nova-api/{resource}/detach                                                                 | nova.api.                   |
| GET|HEAD | nova-api/{resource}/field/{field}                                                          | nova.api.                   |
| GET|HEAD | nova-api/{resource}/filters                                                                | nova.api.                   |
| DELETE   | nova-api/{resource}/force                                                                  | nova.api.                   |
| DELETE   | nova-api/{resource}/lens/{lens}                                                            | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}                                                            | nova.api.                   |
| POST     | nova-api/{resource}/lens/{lens}/action                                                     | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/actions                                                    | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/cards                                                      | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/count                                                      | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/filters                                                    | nova.api.                   |
| DELETE   | nova-api/{resource}/lens/{lens}/force                                                      | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/metrics                                                    | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lens/{lens}/metrics/{metric}                                           | nova.api.                   |
| PUT      | nova-api/{resource}/lens/{lens}/restore                                                    | nova.api.                   |
| GET|HEAD | nova-api/{resource}/lenses                                                                 | nova.api.                   |
| GET|HEAD | nova-api/{resource}/metrics                                                                | nova.api.                   |
| GET|HEAD | nova-api/{resource}/metrics/{metric}                                                       | nova.api.                   |
| GET|HEAD | nova-api/{resource}/morphable/{field}                                                      | nova.api.                   |
| GET|HEAD | nova-api/{resource}/relate-authorization                                                   | nova.api.                   |
| PUT      | nova-api/{resource}/restore                                                                | nova.api.                   |
| GET|HEAD | nova-api/{resource}/soft-deletes                                                           | nova.api.                   |
| POST     | nova-api/{resource}/trix-attachment/{field}                                                | nova.api.                   |
| DELETE   | nova-api/{resource}/trix-attachment/{field}                                                | nova.api.                   |
| DELETE   | nova-api/{resource}/trix-attachment/{field}/{draftId}                                      | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}                                                           | nova.api.                   |
| PUT      | nova-api/{resource}/{resourceId}                                                           | nova.api.                   |
| POST     | nova-api/{resource}/{resourceId}/attach-morphed/{relatedResource}                          | nova.api.                   |
| POST     | nova-api/{resource}/{resourceId}/attach/{relatedResource}                                  | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}/attachable/{field}                                        | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}/download/{field}                                          | nova.api.                   |
| DELETE   | nova-api/{resource}/{resourceId}/field/{field}                                             | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}/metrics/{metric}                                          | nova.api.                   |
| POST     | nova-api/{resource}/{resourceId}/update-attached/{relatedResource}/{relatedResourceId}     | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}/update-fields                                             | nova.api.                   |
| GET|HEAD | nova-api/{resource}/{resourceId}/update-pivot-fields/{relatedResource}/{relatedResourceId} | nova.api.                   |
| DELETE   | nova-api/{resource}/{resourceId}/{relatedResource}/{relatedResourceId}/field/{field}       | nova.api.                   |
| GET|HEAD | nova-vendor/maatwebsite/laravel-nova-excel/download                                        | laravel-nova-excel.download |
| POST     | password/email                                                                             | nova.password.email         |
| GET|HEAD | password/reset                                                                             | nova.password.request       |
| POST     | password/reset                                                                             | nova.                       |
| GET|HEAD | password/reset/{token}                                                                     | nova.password.reset         |
+----------+--------------------------------------------------------------------------------------------+-----------------------------+
```

After route cache

```
art route:list --columns=method,uri,name --name=nova
+----------+-------------------------------------------------------------------+-----------------------------+
| Method   | URI                                                               | Name                        |
+----------+-------------------------------------------------------------------+-----------------------------+
| POST     |                                                                   | nova.login                  |
| GET|HEAD |                                                                   | nova.logout                 |
| POST     | email                                                             | nova.password.email         |
| POST     | nova-api/{resource}/{resourceId}/attach-morphed/{relatedResource} | nova.api.                   |
| GET|HEAD | nova-vendor/maatwebsite/laravel-nova-excel/download               | laravel-nova-excel.download |
| POST     | reset                                                             | nova.                       |
| GET|HEAD | reset                                                             | nova.password.request       |
| GET|HEAD | reset/{token}                                                     | nova.password.reset         |
+----------+-------------------------------------------------------------------+-----------------------------+
```

1. Login and logout route is no longer correct.
2. All `nova.api.` route truncated.
3. `password` prefix got truncated.

```
art route:list --columns=method,uri,name --path=nova-api

+--------+-------------------------------------------------------------------+-----------+
| Method | URI                                                               | Name      |
+--------+-------------------------------------------------------------------+-----------+
| POST   | nova-api/{resource}/{resourceId}/attach-morphed/{relatedResource} | nova.api. |
+--------+-------------------------------------------------------------------+-----------+
```

Reference: https://github.com/laravel/framework/issues/31732

**Comment by jbrooksuk** [comment]
I've put together a fix at https://github.com/laravel/nova/pull/825. I'm not sure about all of the route names at the moment though.

**Comment by crynobone** [comment]
@jbrooksuk That does solved most of the problem, but 

```
+----------+--------------------------------------------------------------------------------------------+-------------------------------------------+
| Method   | URI                                                                                        | Name                                      |
+----------+--------------------------------------------------------------------------------------------+-------------------------------------------+
| GET|HEAD |                                                                                            | nova.login.form                           |
| POST     |                                                                                            | nova.login                                |
| GET|HEAD |                                                                                            | nova.logout                               |
```

these routes missing `login` on URI.

```
| GET|HEAD | reset                                                                                      | nova.password.request                     |
| POST     | reset                                                                                      | nova.password.update                      |
| GET|HEAD | reset/{token}                                                                              | nova.password.reset                       |
+----------+--------------------------------------------------------------------------------------------+-------------------------------------------+
```

And these route missing `password` prefix in the URI. I'm guessing there still some bug in the new route caching

**Comment by jbrooksuk** [comment]
@crynobone those routes are generated by `laravel/ui`.

**Comment by crynobone** [comment]
@jbrooksuk I'm not using `laravel/ui`, `laravel/nova` loading it for my project. So how can I fix this?

Btw in my project I don't have any authentication or password reset other that what's available from Nova.

```php
    
        Nova::routes()
            ->withAuthenticationRoutes()
            ->withPasswordResetRoutes()
            ->register();
```

**Comment by crynobone** [comment]
If I reset the configuration `nova.path` to `/nova` and run the `route:cache`, everything goes back to normal. This only happen when when you set it to `/`.

```
| GET|HEAD | nova/login                                                                                 | nova.login.form                           |
| POST     | nova/login                                                                                 | nova.login                                |
| GET|HEAD | nova/logout                                                                                | nova.logout                               |
| POST     | nova/password/email                                                                        | nova.password.email                       |
| POST     | nova/password/reset                                                                        | nova.password.update                      |
| GET|HEAD | nova/password/reset                                                                        | nova.password.request                     |
| GET|HEAD | nova/password/reset/{token}                                                                | nova.password.reset                       |
+----------+--------------------------------------------------------------------------------------------+-------------------------------------------+
```

**Comment by vadimsg** [comment]
Can confirm that latest Nova doesn't work with route:cache. 

php 7.4
Nova 3.0.1
Laravel 7

**Comment by SuddenlyRust** [comment]
Same problem here. Cannot access the nova backend after I used `route: cache`.

Following happens when I visit  `/nova`:

When I am logged out: **I get too many redirects error when I try to visit** 

When I am logged in: **Following error is thrown**
![image](https://user-images.githubusercontent.com/14092921/76077793-b0df0280-5fa1-11ea-9fcf-709f1b0b78fd.png)

PHP: 7.4.3
Nova: v3.0.1
Laravel: v7.0.4

**Comment by jbrooksuk** [comment]
Does the PR linked above work for you?

**Comment by SuddenlyRust** [comment]
With [pull request 825](https://github.com/laravel/nova/pull/825) it does work fine again.

Only problem is I no longer can directly visit a link. If I use the navigation everything works fine but when I try to call `nova/dashboards/main` directly or press F5 I get an `404` page not found.

**Comment by jbrooksuk** [comment]
Oh? I don't have this issue 🤔

**Comment by jbrooksuk** [comment]
Ignore me, I tested against the wrong project 🙊 

This 404 is because we register the Nova routes in an event listener, which the route cache wouldn't fire. I'll investigate.

**Comment by jbrooksuk** [comment]
Okay, I've pushed a fix which removes the registration of routes within an event listener. If someone could test this again (I've already tested, but just in case...) then I will get this released.

**Comment by SuddenlyRust** [comment]
@jbrooksuk it's working for me now! 👍 nice work

**Comment by mvdnbrk** [comment]
https://laravel.com/docs/7.x/upgrade#unique-route-names

In PR laravel/nova#825 route names like `search`, `dashboard`, `login.form`, `password.update` are used and are very generic.

These named routes may also be used for a front-end app and may collide.

Maybe we should prefix all nova route names with `nova.`

**Comment by jbrooksuk** [comment]
Yeah, the routes are already prefixed. The search route is `nova.api.search`.

**Comment by mvdnbrk** [comment]
@jbrooksuk  laravel/nova#825 seems to fix it 👍

**Comment by jbrooksuk** [comment]
@driesvints is investigating why this has broken. We'll wait for this before we patch.

**Comment by jbrooksuk** [comment]
This is now fixed and released in v3.0.2
